source/common/common/key_value_store_base.cc

```diff
@@ -53,32 +58,54 @@ bool KeyValueStoreBase::parseContents(absl::string_view contents) {
       ENVOY_LOG(warn, error);
       return false;
     }
-    addOrUpdate(key.value(), value.value());
+    addOrUpdate(key.value(), value.value(), absl::nullopt);
   }
   return true;
 }
 
-void KeyValueStoreBase::addOrUpdate(absl::string_view key_view, absl::string_view value_view) {
+void KeyValueStoreBase::addOrUpdate(absl::string_view key_view, absl::string_view value_view,
+                                    absl::optional<std::chrono::seconds> ttl) {
   ENVOY_BUG(!under_iterate_, "addOrUpdate under the stack of iterate");
   std::string key(key_view);
   std::string value(value_view);
+  // Do not add if ttl is <= 0
+  if (ttl && ttl <= std::chrono::seconds(0)) {
+    ASSERT(false);
+    return;
+  }
+
   // Attempt to insert the entry into the store. If it already exists, remove
   // the old entry and insert the new one so it will be in the proper place in
   // the linked list.
   if (!store_.emplace(key, value).second) {
     store_.erase(key);
     store_.emplace(key, value);
+    ttl_manager_.clear(key);
   }
   if (max_entries_ && store_.size() > max_entries_) {
     store_.pop_front();
   }
+  if (ttl) {
+    ttl_manager_.add(std::chrono::milliseconds(ttl.value()), key);
+  }
+  if (!flush_timer_->enabled()) {
+    flush();
+  }
+}
+
+void KeyValueStoreBase::ttlCallback(const std::vector<std::string>& keys) {
+  ENVOY_BUG(!under_iterate_, "remove_batch under the stack of iterate");
```

Comment:
nit: the message here says "remove_batch" but that's not the name of the method.